Profiling: sys.settrace implementation.

malinah opened the pull request:
This is a working concept of profiling API related to issue #3864.

Please review this pull-request and comment especially on the overall integration into the Micropython.

So far the main concern of the profiler is runtime instruction tracing used to create detailed code coverage reports. To achieve this the analogue of [sys.settrace](
https://docs.python.org/3/library/sys.html#sys.settrace) was implemented and puts Micropython closer to feature a debugger #3009.

### Building and Running

So far only the Unix port is concerned.

Enable profiling in ```ports/unix/mpconfigport.h``` like so:
```c
#define MICROPY_PY_SYS_SETTRACE (1)
```

```MICROPY_PERSISTENT_CODE_SAVE``` is required by the ```MICROPY_PY_SYS_TRACE```.

### Python API

#### ```sys.settrace(tracefunc)```
As defined in original Python3 documentation [sys.settrace(tracefunc)](
https://docs.python.org/3/library/sys.html#sys.settrace). The Frame object and Code objects are only partially implemented.

stinos commented:
This looks pretty neat (some commit message will need rewording and more explanation probably), and the example runs fine for the msvc port. I do get a `maximum recursion depth exceeded` in debug builds though but that's expected.

stinos commented on py/modsys.c:
```diff
@@ -147,6 +161,65 @@ STATIC mp_obj_t mp_sys_getsizeof(mp_obj_t obj) {
 #endif
 
+#if MICROPY_PY_SYS_PROFILING
+// prof_mode([mode]): Optionaly set the profiling mode and return the previous.
```
Would be more consistent to add comments in the header file I think

pfalcon commented:
I think this should depend on https://github.com/micropython/micropython/issues/3412 as a vehicle to make this (and many more) feature available in a standard way.

pfalcon commented on py/objattrtuple.c:
```diff
@@ -75,7 +75,11 @@ mp_obj_t mp_obj_new_attrtuple(const qstr *fields, size_t n, const mp_obj_t *item
     o->len = n;
     for (size_t i = 0; i < n; i++) {
-        o->items[i] = items[i];
+        if (items) {
```
This is *slow*!

pfalcon commented:
Has unrelated commits:

> moductypes: add layout parameter to uctypes.sizeof
> modutimeq: add discard method

stinos commented:
> I think this should depend on #3412 as a vehicle to make this (and many more) feature available in a standard way.

You mean a standard way other than defining features in mpconfigport.h? Otherwise these are just orthogonal issues (or i'm missing some point here)

pfalcon commented:
Let's take sys.atexit() as an example. Not related to profiling per se. Should be in a separate commit, with its own config option, then test - everything as usual. Commit message should explain why that name was chosen, given that it's not compatible with CPython.

And likely, submitted as a separate PR for quicker merging (and unloading this PR for noise-free review).

pfalcon commented:
> You mean a standard way other than defining features in mpconfigport.h? Otherwise these are just orthogonal issues (or i'm missing some point here)

I mean that one argument against adding more and more features was "this is not useful for intended [production] usage of MicroPython (fitting it where other stuff doesn't fit)" and "this can only lead to fragmentation".

So, IMHO, before going to stuff more and more features, answers to those concerns should be given, and https://github.com/micropython/micropython/issues/3412 tries to do just that.

stinos commented:
Yeah I get that, it's the *how* I don't see. Like, practically, what does this entail? Just `make special_dev_build` which configures PP directives/extra source files or something completely different? But probably that's a discussion which should be held on that issue instead of here :P

pfalcon commented:
Sure, nothing magic: https://github.com/pfalcon/micropython/commit/1752b8a10a4bb39d97f3840cd5d5250952de3b8f

pfalcon commented:
@malinah : Oh, and otherwise, this is a great feature of course, thanks for posting!

pfalcon commented on py/mpstate.h:
```diff
@@ -220,6 +227,12 @@ typedef struct _mp_state_thread_t {
 
+    #if MICROPY_PY_SYS_PROFILING
+    mp_obj_t prof_instr_tick_callback;
+    bool prof_instr_tick_callback_is_executing;
```
bool should go somewhere in the structure hole. (with a comment here)

pfalcon commented on py/mpstate.h:
```diff
@@ -220,6 +227,12 @@ typedef struct _mp_state_thread_t {
     mp_obj_dict_t *dict_globals;
 
+    #if MICROPY_PY_SYS_PROFILING
+    mp_obj_t prof_instr_tick_callback;
```
"prof_" should be dropped from these, not related to profile. Compare to what I had from my experiments with tracing JIT:

~~~
extern mp_code_state_t *mp_current_code_state;
~~~

pfalcon commented on py/modsys.c:
```diff
@@ -147,6 +161,65 @@ STATIC mp_obj_t mp_sys_getsizeof(mp_obj_t obj) {
+STATIC mp_obj_t mp_sys_settrace(mp_obj_t obj) {
+
+    if (mp_obj_is_callable(obj)) {
+        MP_STATE_THREAD(prof_instr_tick_callback) = obj;
```
How this variable got named "instr_**tick**_callback", if it appears in function named "settrace"? ;-)

pfalcon commented:
@dpgeorge , You may want to review, and if correct, merge commits "tools/mpy-tools.py: Fix opcode size enumeration and constant definition." and "py/bc.c: Fix opcode size enumeration." ahead of anything else. Then setup suitable tests to avoid that in the future.

malinah commented:
@pfalcon I see we are working around the same area (bytecode, vm) - https://github.com/pfalcon/micropython/commits/pfalcon vs https://github.com/trezor/micropython/commits/malinah/profiling. We should sync to avoid redundant work. Where can I reach you? Cheers.

pfalcon commented:
> @pfalcon I see we are working around the same area (bytecode, vm). 
> We should sync to avoid redundant work.

Sure, that's why comments above and https://github.com/micropython/micropython/pull/4342.

> Where can I reach you?

I'm here.

malinah commented:
Hi, I reworked the whole implementation and it should make more sense now. However, I have a problem building the uPy with **MICROPY_PERSISTENT_CODE_SAVE** enabled. I'm getting some errors in *py/persistentcode.c* that I'm currently ugly-patching (not in this PR) and I don't know how to fix them properly. Can you give me a feedback please? Cheers and thanks.

pfalcon commented on py/mpstate.h:
```diff
@@ -245,6 +250,14 @@ typedef struct _mp_state_thread_t {
+    bool prof_callback_is_executing;
+    #endif // MICROPY_PY_SYS_TRACE
+    #if MICROPY_ACCESS_CODE_STATE
+    struct _mp_code_state_t *prof_code_state;
```
It's unclear what was reworked, as one of the basic comments was that reference to the current code state being executed isn't really related to profiling per se, so shouldn't have "prof_" prefix.

malinah replied:
What was reworked is the tracing integration inside the *py/vm.c* and settrace callback invocation with proper frame and code objects. I renamed the pointer. It should be fixed now.

pfalcon commented on py/bc.h:
```diff
@@ -84,6 +84,10 @@ typedef struct _mp_code_state_t {
     struct _mp_code_state_t *prev;
     #endif
+    #if MICROPY_PY_SYS_TRACE
+    struct _mp_code_state_t *prev_state;
```
There's already `struct _mp_code_state_t *prev;` above.

malinah replied:
The **prev_state** is *MICROPY_STACKLESS* agnostic which **prev** is not. I tried to use the **prev** but it was difficult to change the logic around it to separate the normal execution traceback from settrace callback execution which is needed for the **frame** object. In this sense they are not the same. Also this feature is not mature so it seems easier to keep it's members separate from the uPy core (easier to debug).

pfalcon commented on py/objfun.h:
```diff
@@ -28,11 +28,19 @@
     const mp_uint_t *const_table;   // constant table
+#if MICROPY_PY_SYS_TRACE
+    const mp_raw_code_t *rc;
+    mp_obj_t line_of_definition;
```
This is clearly superfluous, line info is already stored with bytecode object.

malinah replied:
In the case a function was NOT defined at compile time this variable is needed; e.g. *eval*.

malinah replied:
From what I understand the bytecode contains the line info of the first opcode of the first expression INSIDE the block. Not the line where the block was created; e.g. *MP_BC_MAKE_FUNCTION*.

```Python
283:    def do(): # MP_BC_MAKE_FUNCTION Somehwere inside the bytecode `AAA`
284:        
285:        print("Who loves the sun?") # bytecode `BBB` starts here
...
301:    do()
```

So when there is a bytecode execution switch from `AAA` to `BBB` I need the line number of the **def** in order to trigger the *call* trace event with the correct line numbering.

Trace event *call* is triggered at line 301 and jumps to line 285 but the called function is defined at line 283. This information needs to be somehow propagated. Hence the `line_of_definition`. If you have another method please tell me.

pfalcon replied:
> From what I understand the bytecode contains the line info of the first opcode of the first expression INSIDE the block.

Ok, if the difference is important for your usecase, please add code comment explaining this.

pfalcon reviewed (changes requested):
From 1st look, the update spots about the same issues as the original patch had.

pfalcon commented on py/emitglue.h:
```diff
@@ -85,6 +85,8 @@ typedef struct _mp_raw_code_t {
     #if MICROPY_PY_SYS_TRACE
     mp_bytecode_prelude_t prelude;
+    // Python source line where the raw_code was created e.g. MP_BC_MAKE_FUNCTION
+    // Required by the `call` trace event.
```
"... This is different from lineno info stored in prelude, which provides line number for first statement of a function. Required to properly implement "call" trace event."

(Adjust to provide actual info, the point that it should explain why seemingly duplicate data is stored.)

malinah commented:
Hi, If you could give me more feedback please that would be great. @dpgeorge @pfalcon @stinos. Thank you and cheers!

stinos commented:
From a quick look: code itself looks sort of fine and operational, but it's a bit hard to provide constructive feedback on it because the commits look like work in progress built on top of work in progress. I'd suggest to rework this so it's close to it's final state:
- remove dead code like commented out code, sections guarded with `if 0` without explanation why, bunch of TODO comments etc. Also e.g. in your tests: all your functions are called `test_xxx` so it is really unnecessary to have a comment `# xxx` on the line above them.
- rebase the whole thing, squash and/or split fixup commits where they belong; you have commits like 'Fix off by one', I get that it's nice for you to remember but this cannot be merged as a separate commit. Likewise for commits which consist only of renaming of variables. And as mentioned earlier, introducing new features which aren't strictly related to or requires for settrace like `atexit` should be in separate commits.
- on the other hand I'm not sure it's interesting to have separate commits for adding new files and actually using them (just my idea, others might disagree), it's pretty hard to tell now from which commit on there's actually a usable implementation, e.g. there's a separate commit with 'py: Stitch tracing into the vm.' but is there any use for the previous commits if this code isn't present?
- add proper commit messages: most commits consists of only a subject line, general guideline for commits is that the commit message body explains why a change is needed and what it does. Not saying all commits need this but just as an example there's a commit 'unix: Trigger uatxit callback when actually exiting.' Looks interesting, but to understand why this is suddenly needed from the code alone it looks like  I'd have to go through the code myself. (i.e. MICROPY_PY_SYS_UATEXIT is implemented already, does it not work as expected now?)  If the commit would explain why this is needed and what the effect is everything is clear at first sight..

prusnak commented:
@stinos While I agree with the comments above, it is possible to view squashed changes as one commit here: https://github.com/micropython/micropython/pull/4265/files

stinos commented:
@prusnak thanks but I'm aware of that already :)

dpgeorge commented:
@malinah thanks a lot for posting this PR, it's a welcome feature for sure, useful for (at least) profiling, debugging, coverage testing.

The commits here still need a fair bit of work to be in a state for merging.  There are minor things like camel-case variable names and duplicated code (eg `prof_extract_prelude` functionality exists in `persistentcode.c`), and lots of seemingly unused code (eg `prof_bytecode_linenox` and `PROF_PRINT_INSTR`).

But more importantly, from a bigger picture point of view: ideally the commits would be made to add logically independent functionality.  Eg first just add bare-bones `sys.settrace()` functionality that just calls the callback with `None` passed in.  Then move on to implement frame objects, etc, gradually adding functionality.  It doesn't have to be 1 commit for each set of working functionality, but it would be great if there were a handful of commits that added something self contained (even better with working tests), which could be more easily reviewed and merged.

Anyway, I'm not sure how you'd like to proceed because it will be a lot more work to make it a clean PR in this way.  At the very least please remove unused code so there's less to review.

----

I did test this PR on unix and stm32 (with minor changes to get it working on stm32) and it does work very well, and has a very faithful API the same as CPython.  So well done there!

On unix 64-bit the **code size increase is around 11k**.  On stm32 it's **around 8.5k**.  That's quite acceptable IMO.

As for performance, I tested both these ports before and after enabling the profiling feature.  Note that with profiling enabled it doesn't actually set `sys.settrace()`, the following numbers are testing the VM speed with the additional hooks needed by the profiler (overhead of profiling when not used).

For unix 64-bit:
```
diff of scores (higher is better)
N=5000 M=1000             unix-base ->  unix-prof         diff      diff% (error%)
bm_chaos.py                14832.12 ->   11159.97 :   -3672.15 = -24.758% (+/-7.65%)
bm_fannkuch.py                 7.33 ->       5.49 :      -1.84 = -25.102% (+/-1.84%)
bm_float.py               301922.69 ->  220679.32 :  -81243.37 = -26.909% (+/-3.23%)
bm_hexiom.py                 673.90 ->     446.06 :    -227.84 = -33.809% (+/-2.47%)
bm_nqueens.py             300178.77 ->  222071.98 :  -78106.79 = -26.020% (+/-3.16%)
bm_pidigits.py              2055.16 ->    2028.26 :     -26.90 =  -1.309% (+/-2.46%)
misc_aes.py                17334.31 ->   10074.71 :   -7259.60 = -41.880% (+/-3.77%)
misc_mandel.py            128709.37 ->  104342.61 :  -24366.76 = -18.932% (+/-4.67%)
misc_pystone.py            91528.72 ->   62108.57 :  -29420.15 = -32.143% (+/-3.61%)
misc_raytrace.py           25008.10 ->   19785.31 :   -5222.79 = -20.884% (+/-4.55%)
viper_call0.py             24579.51 ->   23262.08 :   -1317.43 =  -5.360% (+/-11.54%)
viper_call1a.py            23672.42 ->   23459.22 :    -213.20 =  -0.901% (+/-13.28%)
viper_call1b.py            18712.93 ->   19400.68 :    +687.75 =  +3.675% (+/-13.61%)
viper_call1c.py            20015.56 ->   18945.96 :   -1069.60 =  -5.344% (+/-9.03%)
viper_call2a.py            22129.29 ->   23672.68 :   +1543.39 =  +6.974% (+/-13.03%)
viper_call2b.py            16035.62 ->   17589.49 :   +1553.87 =  +9.690% (+/-13.30%)
```

For PYBv1.0:
```
diff of scores (higher is better)
N=100 M=100               pyb-base  ->   pyb-prof         diff      diff% (error%)
bm_chaos.py                  294.77 ->     259.99 :     -34.78 = -11.799% (+/-0.01%)
bm_fannkuch.py                72.11 ->      62.56 :      -9.55 = -13.244% (+/-0.00%)
bm_float.py                 4779.11 ->    4204.22 :    -574.89 = -12.029% (+/-0.01%)
bm_hexiom.py                  34.46 ->      27.34 :      -7.12 = -20.662% (+/-0.01%)
bm_nqueens.py               4103.28 ->    3374.21 :    -729.07 = -17.768% (+/-0.01%)
bm_pidigits.py               624.21 ->     596.84 :     -27.37 =  -4.385% (+/-0.36%)
misc_aes.py                  342.47 ->     280.31 :     -62.16 = -18.150% (+/-0.00%)
misc_mandel.py              2980.08 ->    2672.15 :    -307.93 = -10.333% (+/-0.01%)
misc_pystone.py             1875.17 ->    1513.70 :    -361.47 = -19.277% (+/-0.00%)
misc_raytrace.py             297.19 ->     265.28 :     -31.91 = -10.737% (+/-0.01%)
viper_call0.py               576.83 ->     576.70 :      -0.13 =  -0.023% (+/-0.00%)
viper_call1a.py              550.37 ->     550.25 :      -0.12 =  -0.022% (+/-0.00%)
viper_call1b.py              434.90 ->     434.85 :      -0.05 =  -0.011% (+/-0.01%)
viper_call1c.py              439.47 ->     439.37 :      -0.10 =  -0.023% (+/-0.01%)
viper_call2a.py              536.28 ->     536.19 :      -0.09 =  -0.017% (+/-0.01%)
viper_call2b.py              377.28 ->     377.22 :      -0.06 =  -0.016% (+/-0.00%)
```

Unix sees a loss of 20-40% performance.  PYBv1.0 sees a 10-20% loss.  They both agree that `misc_aes`, `bm_hexiom` and `misc_pystone` are the tests with the biggest performance loss (they have lots of lines of code so makes sense that VM performance is down).  And `bm_pidigits` sees very little loss because it's mainly stressing bignum arithmetic (makes sense).

dpgeorge commented on ports/unix/main.c:
```diff
@@ -642,6 +642,14 @@ MP_NOINLINE int main_(int argc, char **argv) {
         }
     }
 
+    #if MICROPY_PY_SYS_UATEXIT
```
The whole `sys.uatexit()` functionality needs to be in a completely separate PR, it's not needed for `sys.settrace()` (as far as I can tell).

dpgeorge commented on ports/unix/mpconfigport.h:
```diff
@@ -90,6 +90,13 @@
+#define MICROPY_PY_SYS_TRACE (0)
+#if MICROPY_PY_SYS_TRACE
+#define MICROPY_PY_SYS_UATEXIT (1)
+#define MICROPY_ACCESS_CODE_STATE (1)
```
Why is `MICROPY_ACCESS_CODE_STATE` a separate config option?  It looks like it can't be enabled on its own, and `SYS_TRACE` requires it.

malinah replied:
Somehow I understood that it is desirable to have it as an independent feature #4342 .

dpgeorge replied:
I see.  But it looks like the `current_code_state` variable is not restored on `FRAME_LEAVE` unless `SYS_TRACE` is also enabled, so this feature won't work by itself.  Anyway, I'd prefer to keep this PR as minimal as possible and not have extra configs like this.  They can be added later on if they prove useful.

malinah replied:
Awesome, I agree.

dpgeorge commented on py/emitglue.c:
```diff
@@ -74,6 +78,27 @@ void mp_emit_glue_assign_bytecode(mp_raw_code_t *rc, const byte *code,
+    mp_bytecode_prelude_t *prelude = &rc->prelude;
+    prof_extract_prelude(code, prelude);
+    
+    #if 0
```
Please remove unused code.

dpgeorge commented on py/emitglue.c:
```diff
@@ -82,7 +107,7 @@ void mp_emit_glue_assign_bytecode(mp_raw_code_t *rc, const byte *code,
 #endif
 #if MICROPY_DEBUG_PRINTERS
-    if (mp_verbose_flag >= 2) {
+    if (mp_verbose_flag >= 2 || 0) {
```
Unnecessary change.

dpgeorge commented on py/emitglue.h:
```diff
@@ -63,6 +83,14 @@ typedef struct _mp_raw_code_t {
     uint16_t n_obj;
     uint16_t n_raw_code;
+    #if MICROPY_PY_SYS_TRACE
+    mp_bytecode_prelude_t prelude;
```
Why does this prelude need to be store explicitly?  Is it just for ease-of-access of the state elements (instead of decoding it all the time)?

malinah replied:
The pointer (line_info) to the *line number info* section is needed in every callback invocation and every frame change. Also (qstr_source_file) *source_file* and (qstr_block_name) *simple_name* are often needed. Decoding it all the time is quite a slowdown. Especially for LINE events and in the future for INSTR events. Also in the future it may be desirable to precompute all bytecode metada (*line number info*, etc.) in to some faster read format. But for now I think the current tradeoff is OK.

dpgeorge commented on py/modsys.c:
```diff
@@ -180,6 +209,14 @@ STATIC const mp_rom_map_elem_t mp_module_sys_globals_table[] = {
+    { MP_ROM_QSTR(MP_QSTR_uatexit), MP_ROM_PTR(&mp_sys_uatexit_obj) },
+    #endif
+
+    #if MICROPY_PY_SYS_TRACE
```
Since the function is `settrace` the config option should really be `MICROPY_PY_SYS_SETTRACE` (to be consistent with the other options).

dpgeorge commented on py/profiling.c:
```diff
@@ -0,0 +1,1046 @@
+#include "py/bc0.h"
+#include "py/gc.h"
+
+#if PROF_PRINT_INSTR
```
This whole section looks like unused code.

malinah replied:
The PROF_PRINT_INSTR symbol is enabled on-demand in py/profiling.h . It is intended for debugging purposes only.

dpgeorge replied:
Do you mean debugging the settrace feature?  It's a lot of code just for debugging, and also duplicates things from vm.c/showbc.c.  It would be good to at least have the addition of this debugging code as a separate commit, so it's easier to understand what's needed/not needed for the basic settrace implementation.

dpgeorge commented on py/profiling.c:
```diff
@@ -0,0 +1,1046 @@
+    return source_line;
+}
+
+uint prof_bytecode_linenox(const byte *line_info, size_t bc) {
```
This looks unused.

dpgeorge commented on py/profiling.c:
```diff
@@ -0,0 +1,1046 @@
+    return source_line;
+}
+
+void prof_extract_prelude(const byte *bytecode, mp_bytecode_prelude_t *prelude) {
```
Is it possible to reuse `extract_prelude` from `persistentcode.c`?

malinah replied:
It is not possible to reuse the `extract_prelude` as it is right now. It would have to be changed and I would like to keep this PR as decoupled as possible while in the WIP mode. I would make such a change in a later stage if it will make sense.

dpgeorge replied:
Ok, makes sense.

dpgeorge commented on py/profiling.c:
```diff
@@ -0,0 +1,1046 @@
+
+
+//
+// CODE object
```
For comments that separate sections, please use:
```c
/******************************************************************************/
// code object
```

dpgeorge commented on py/profiling.c:
```diff
@@ -0,0 +1,1046 @@
+    // }
+
+    // start = prelude->n_pos_args + prelude->n_kwonly_args + rc->n_obj;
+    // stop  = prelude->n_pos_args + prelude->n_kwonly_args + rc->n_obj + rc->n_raw_code;
```
Is this code needed/useful?  If not please remove it.

dpgeorge commented on py/profiling.c:
```diff
@@ -0,0 +1,1046 @@
+            if (buffer_index + 2 > buffer_size) {
+                // mp_printf(&mp_plat_print, "has to realloc lnotab\n");
+                #if MICROPY_MALLOC_USES_ALLOCATED_SIZE
+                buffer = m_realloc(buffer, buffer_size, buffer_size + buffer_chunk_size);
```
You should be able to use `m_renew(byte, buffer, buffer_size, buffer_size + buffer_chunk_size)`, then no need for a #if.

dpgeorge commented on py/profiling.c:
```diff
@@ -0,0 +1,1046 @@
+
+    // mp_printf(&mp_plat_print, "lnotab size %d overshoot %d\n", buffer_index, buffer_size - buffer_index);
+    mp_obj_t o = mp_obj_new_bytes(buffer, buffer_index);
+    m_free(buffer, buffer_size);
```
This should be `m_del(byte, buffer, buffer_size)`

dpgeorge commented on py/profiling.c:
```diff
@@ -0,0 +1,1046 @@
+    }
+
+    if (*code_state->ip == MP_BC_FOR_ITER) {
+        // assert(false);
```
What happens if it's a FOR_ITER, is this unimplemented so far?

dpgeorge commented on py/profiling.c:
```diff
@@ -0,0 +1,1046 @@
+    return top;
+}
+
+mp_obj_t prof_instr_tick(mp_code_state_t *code_state, bool isException) {
```
`isException` should be `is_exception` (or just `is_exc`).

dpgeorge commented on py/profiling.c:
```diff
@@ -0,0 +1,1046 @@
+    }
+
+    // OPCODE
+    if (false) {
```
A way to implement this would be to support setting `frame.f_trace_opcodes=True`, a feature added in CPython 3.7.

dpgeorge commented on py/profiling.h:
```diff
@@ -0,0 +1,108 @@
+    bool trace_opcodes;
+} mp_obj_frame_t;
+
+typedef struct __attribute__((__packed__)) {
```
Please don't use this packed attribute, it's not needed.

dpgeorge commented on py/showbc.c:
```diff
@@ -540,7 +540,7 @@ const byte *mp_bytecode_print_str(const byte *ip) {
                 printf("BINARY_OP " UINT_FMT " %s", op, qstr_str(mp_binary_op_method_name[op]));
             } else {
-                printf("code %p, byte code 0x%02x not implemented\n", ip, ip[-1]);
+                printf("code %p, byte code 0x%02x not implemented\n", ip-1, ip[-1]);
```
This can be fixed separately to this PR.

dpgeorge commented on py/vm.c:
```diff
@@ -148,7 +208,46 @@ mp_vm_return_kind_t mp_execute_bytecode(mp_code_state_t *code_state, volatile mp
+#if MICROPY_PY_SYS_TRACE
+// Trace CALL event.
+if (!prof_is_executing && prof_trace_cb) {
+
```
The following chunk of code can probably be factored to a function in `profiling.c`, then `prof_callback_invoke()` can be made private to `profiling.c`.

dpgeorge commented:
@malinah I see you force-pushed this PR.  Please let me know when it's ready for another review.

malinah committed:
py/profiling: Initial implementation of `sys.settrace`.

malinah committed:
py: Integrate `sys.tracing` into the project.

 - Introducing the feature into the vm.c and surrounding runtime.
 - Introducing the feature into the build system and configuration.
 - To enable the feature enable MICROPY_PY_SYS_SETTRACE in the
   mpconfigport.h in the unix port. Other ports are not supported
   right now.

malinah committed:
tests: Integrate tests for `sys.settrace` feature.

 - tests/misc: Added initial tests for `sys.settrace` feature.
 - tests/run-tests: register misc/trace_profilin.py as special testcase.
 - tests/cmdline/cmd_showbc.py had to be updated because of the changes
   in the compile.c regarding the `sys.settrace` specific line emitters.

malinah committed:
py/profiling: Added `sys.settrace` feature debugging.

malinah commented:
> @malinah I see you force-pushed this PR. Please let me know when it's ready for another review.

@dpgeorge I think I modified the code and commits according to resolved comments.  Check it out. Cheers!

dpgeorge commented on py/compile.c:
```diff
@@ -800,6 +800,9 @@ STATIC qstr compile_classdef_helper(compiler_t *comp, mp_parse_node_struct_t *pn
 
     // compile the class
+#if MICROPY_PY_SYS_SETTRACE
+    EMIT_ARG(set_source_line, pns->source_line);
```
It might make sense to enable these 4 set-source-line emits unconditionally.  I assume they are there to get better line numbering for these cases, is that correct?  Wouldn't such number be useful regardless of settrace being enabled?

malinah replied:
> It might make sense to enable these 4 set-source-line emits unconditionally.
> Wouldn't such number be useful regardless of settrace being enabled?

Nobody was missing them so far. Unconditional emitters would make the _release_ frozen uPy images larger with _useless information_. On Unix port this doesn't matter but for firmware images I guess it does. I didn't make the measurements though. And I'm not aware of any usefulness outside of `sys.settrace`.

> I assume they are there to get better line numbering for these cases, is that correct? 

Yes. The emitters were added in order to get precision line numbering at places where it was missing -- compared to CPython `sys.settrace` output.

dpgeorge replied:
> Nobody was missing them so far.

True!



> Unconditional emitters would make the _release_ frozen uPy images larger with _useless information_.

It's possible to compile bytecode with optimisation which doesn't store the line number table.

For this particular line here I couldn't see any difference in generated bytecode with it included.  Do you have a test case where it makes a difference?

malinah replied:
> For this particular line here I couldn't see any difference in generated bytecode with it included. Do you have a test case where it makes a difference? 

Unfortunately I'm not able to reproduce the testcase for this line. It will be easier to remove it for now and wait for the issue to pop up.

dpgeorge replied:
Sounds good.

dpgeorge commented on py/mpconfig.h:
```diff
@@ -1548,4 +1553,14 @@ typedef double mp_float_t;
+#error "MICROPY_PY_SYS_SETTRACE requires MICROPY_PERSISTENT_CODE_SAVE to be enabled!"
+#endif
+#if MICROPY_COMP_CONST
+#error "MICROPY_PY_SYS_SETTRACE requires MICROPY_COMP_CONST to be disabled!"
```
Why does it require COMP_CONST to be disabled?  Most ports have this feature enabled so it'd be good to make it work with settrace (not saying that it should be done now, just wondering what the problem is having them both enabled).

malinah replied:
It works with COMP_CONST enabled but the results are not precise -- compared to CPython. The uPy compiler optimisation enabled by COMP_CONST makes some lines _invisible_ eg.
```python
1. from micropython import const
2. _CNT02 = const(123)
```
Line 2 in the above example would be optimised out which is bad for the code coverage tool built upon this feature.

Changing the `error` **requirement** of `MICROPY_COMP_CONST` to `warning` **suggestion** would be more correct way to go.

Something like:
```c
#warning "MICROPY_PY_SYS_SETTRACE suggests MICROPY_COMP_CONST to be disabled for accurate traces.
```

dpgeorge replied:
I see.  Maybe when settrace is enabled the emitter can emit a dummy operation in place here, like LOAD_CONST_NONE + POP_TOP.

malinah replied:
> Maybe when settrace is enabled the emitter can emit a dummy operation in place here, like LOAD_CONST_NONE + POP_TOP.

I'm afraid that would create confusion when the OPCODE tracing will be supported.

Another way (I didn't test it) is to change all `#if MICROPY_COMP_CONST` to `#if MICROPY_COMP_CONST && ! MICROPY_PY_SYS_SETTRACE` in _parse.c_.

If any auto-magic :dizzy: is desired I think this would be the one that would make anybody less angry about unexpected behaviour :smile: Anyway any auto-magic in this case creates coupling and doesn't keep the logic of enabled features explicit.

dpgeorge replied:
Ok, let's leave it as you have it.  The point is (as you mention above) that to get precise results the "x=const(123)" feature must be turned off.

dpgeorge commented on py/compile.c:
```diff
@@ -1633,6 +1636,9 @@ STATIC void compile_try_except(compiler_t *comp, mp_parse_node_t pn_body, int n_
         }
 
+#if MICROPY_PY_SYS_SETTRACE
+        EMIT_ARG(set_source_line, pns_except->source_line);
```
I see how this is useful for settrace.  For the following code:
```python
try:
    raise 1
except Exceptio as er:
    pass
```
It adds the line number info for the "except Exceptio as er:" line, which in this case gives better error reporting when the lookup of `Exceptio` fails.

Actually it'd be better if it were moved up to line 1614.

dpgeorge commented on py/compile.c:
```diff
@@ -3112,6 +3118,9 @@ STATIC void compile_scope(compiler_t *comp, scope_t *scope, pass_kind_t pass) {
         }
 
+#if MICROPY_PY_SYS_SETTRACE
+        EMIT_ARG(set_source_line, pns->source_line);
```
This one is useful in general (not just for settrace) because it'll give much better line number reporting in comprehensions.  Currently we get "line 1" for comprehensions.

malinah replied:
Should I just remove the `#if` and keep it in this PR? Or separate PR?

dpgeorge replied:
I know it's just one line but I think it's a good candidate for a separate commit.  I'm happy to do the commit, or feel free to make a separate PR.

dpgeorge commented on py/compile.c:
```diff
@@ -3151,6 +3160,9 @@ STATIC void compile_scope(compiler_t *comp, scope_t *scope, pass_kind_t pass) {
         }
 
+#if MICROPY_PY_SYS_SETTRACE
+        EMIT_ARG(set_source_line, pns->source_line);
```
I can see this makes a difference to generated bytecode, and only useful for settrace.  So good as-is.

malinah commented:
@dpgeorge All comments worked in and rebased on the current master.

dpgeorge commented:
Superseded by #5026
